core/vm/contracts.go

```diff
@@ -1353,3 +1358,91 @@ func (c *p256Verify) Run(input []byte) ([]byte, error) {
 		return nil, nil
 	}
 }
+
+// gpgEd25519Verify implements native verification for ed25519 signatures produced via gpg
+type gpgEd25519Verify struct{}
+
+var (
+	errMessageTooShort    = errors.New("message too short")
+	errPubKeyTooShort     = errors.New("public key too short")
+	errSignatureTooShort  = errors.New("signature too short")
+	errInvalidPublicKey   = errors.New("invalid public key format")
+	errInvalidSignature   = errors.New("invalid signature format")
+	errVerificationFailed = errors.New("signature verification failed")
+)
+
+// RequiredGas returns the gas required to execute the pre-compiled contract
+func (c *gpgEd25519Verify) RequiredGas(input []byte) uint64 {
+	// You can adjust this value based on your needs
+	return params.GpgEd25519VerifyGas
+}
+
+// Run performs ed25519 signature verification
+func (c *gpgEd25519Verify) Run(input []byte) ([]byte, error) {
+	// Input should be: message_len (32 bytes) || message || pubkey_len (32 bytes) || pubkey || sig_len (32 bytes) || signature
+	if len(input) < 96 { // minimum length for the three length fields
+		return nil, errMessageTooShort
+	}
+
+	// Extract message length and message
+	msgLen := new(big.Int).SetBytes(input[:32]).Uint64()
+	if len(input) < 32+int(msgLen) {
+		return nil, errMessageTooShort
+	}
+	message := input[32 : 32+msgLen]
+
+	// Extract public key length and public key
+	offset := 32 + msgLen
+	pubKeyLen := new(big.Int).SetBytes(input[offset : offset+32]).Uint64()
+	if len(input) < int(offset+32+pubKeyLen) {
+		return nil, errPubKeyTooShort
+	}
+	pubKey := input[offset+32 : offset+32+pubKeyLen]
```

Comment:
Is there a specific length the pubKey is supposed to be that we should verify?

Reply:
Similar to the [comment on signature length](https://github.com/teaxyz/tea-geth/pull/2#discussion_r1897890870). The _armored_ public key length is non-deterministic (exported via `gpg --export --armor <your-key-id>`).

Reply:
Agreed, no way to do this. I'm going to leave this comment open as a reminder that I need to talk to the Tea team to figure out a reasonable upper bound to make sure that (a) there is sufficient gas to handle it, and (b) that there is no DOS risk based on whatever gas params we choose.